File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/AppShakeWarningModels.kt
package com.pillyliu.pinprofandroid

import android.os.VibrationEffect
import android.os.Vibrator
import androidx.compose.ui.graphics.Color

internal enum class AppShakeWarningLevel(
    val title: String,
    val subtitle: String,
    val bundledArtAssetPath: String,
    val tint: Color,
    val glow: Color,
    val displayDurationMillis: Long,
    val hapticStartDelayMillis: Long,
) {
    Danger(
        title = "DANGER",
        subtitle = "A little restraint, if you please.",
        bundledArtAssetPath = "shake-warnings/professor-danger_1024.webp",
        tint = Color(0xFFFF9E2E),
        glow = Color(0xFFFFD15C),
        displayDurationMillis = 3_000L,
        hapticStartDelayMillis = 50L,
    ),
    DoubleDanger(
        title = "DANGER DANGER",
        subtitle = "Really, this is most uncivilised shaking.",
        bundledArtAssetPath = "shake-warnings/professor-danger-danger_1024.webp",
        tint = Color(0xFFFF5729),
        glow = Color(0xFFFF852E),
        displayDurationMillis = 3_500L,
        hapticStartDelayMillis = 200L,
    ),
    Tilt(
        title = "TILT",
        subtitle = "That is quite enough! I will not tolerate any further indignity in this cabinet of higher learning.",
        bundledArtAssetPath = "shake-warnings/professor-tilt_1024.webp",
        tint = Color(0xFFFF2424),
        glow = Color(0xFFFF472E),
        displayDurationMillis = 4_500L,
        hapticStartDelayMillis = 200L,
    ),
}

internal object AppShakeMotionTuning {
    const val sampleRateHz = 30
    const val minimumAcceptedShakeIntervalMillis = 850L
    const val candidateWindowMillis = 180L
    const val strongMagnitudeThreshold = 2.45f
    const val combinedMagnitudeThreshold = 1.85f
    const val combinedPeakAxisThreshold = 1.35f
}

internal fun AppShakeWarningLevel.vibrationEffect(vibrator: Vibrator?): VibrationEffect {
    val supportsAmplitude = vibrator?.hasAmplitudeControl() == true
    val dangerAmplitude = if (supportsAmplitude) 189 else VibrationEffect.DEFAULT_AMPLITUDE
    val doubleDangerAmplitude = if (supportsAmplitude) 209 else VibrationEffect.DEFAULT_AMPLITUDE
    val tiltAmplitude = if (supportsAmplitude) 255 else VibrationEffect.DEFAULT_AMPLITUDE
    return when (this) {
        AppShakeWarningLevel.Danger -> {
            VibrationEffect.createWaveform(
                longArrayOf(0L, 110L),
                intArrayOf(0, dangerAmplitude),
                -1,
            )
        }
        AppShakeWarningLevel.DoubleDanger -> {
            VibrationEffect.createWaveform(
                longArrayOf(0L, 110L, 170L, 110L),
                intArrayOf(0, doubleDangerAmplitude, 0, doubleDangerAmplitude),
                -1,
            )
        }
        AppShakeWarningLevel.Tilt -> {
            VibrationEffect.createWaveform(
                longArrayOf(0L, 140L, 150L, 140L, 150L, 140L),
                intArrayOf(0, tiltAmplitude, 0, tiltAmplitude, 0, tiltAmplitude),
                -1,
            )
        }
    }
}

File: Pinball App 2/Pinball App 2/app/AppShakeWarningModels.swift
import SwiftUI

enum AppShakeWarningLevel: Int {
    case danger = 1
    case doubleDanger = 2
    case tilt = 3

    var title: String {
        switch self {
        case .danger:
            return "DANGER"
        case .doubleDanger:
            return "DANGER DANGER"
        case .tilt:
            return "TILT"
        }
    }

    var subtitle: String {
        switch self {
        case .danger:
            return "A little restraint, if you please."
        case .doubleDanger:
            return "Really, this is most uncivilised shaking."
        case .tilt:
            return "That is quite enough! I will not tolerate any further indignity in this cabinet of higher learning."
        }
    }

    var artAssetName: String {
        switch self {
        case .danger:
            return "ProfessorShakeDanger"
        case .doubleDanger:
            return "ProfessorShakeDoubleDanger"
        case .tilt:
            return "ProfessorShakeTilt"
        }
    }

    var bundledArtFileName: String {
        switch self {
        case .danger:
            return "professor-danger_1024.webp"
        case .doubleDanger:
            return "professor-danger-danger_1024.webp"
        case .tilt:
            return "professor-tilt_1024.webp"
        }
    }

    var tint: Color {
        switch self {
        case .danger:
            return Color(red: 1.00, green: 0.62, blue: 0.18)
        case .doubleDanger:
            return Color(red: 1.00, green: 0.34, blue: 0.16)
        case .tilt:
            return Color(red: 1.00, green: 0.14, blue: 0.14)
        }
    }

    var glow: Color {
        switch self {
        case .danger:
            return Color(red: 1.00, green: 0.82, blue: 0.36)
        case .doubleDanger:
            return Color(red: 1.00, green: 0.52, blue: 0.18)
        case .tilt:
            return Color(red: 1.00, green: 0.28, blue: 0.18)
        }
    }

    var displayDurationNanoseconds: UInt64 {
        switch self {
        case .danger:
            return 3_000_000_000
        case .doubleDanger:
            return 3_500_000_000
        case .tilt:
            return 4_500_000_000
        }
    }

    var hapticStartDelayNanoseconds: UInt64 {
        switch self {
        case .danger:
            return 50_000_000
        case .doubleDanger, .tilt:
            return 200_000_000
        }
    }
}
